### PR TITLE
Recognize multiple parameters in a command, and support them for find, findOne, insertMany

### PR DESCRIPTION
**Pull request by PrashanthCorp**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -718,7 +718,7 @@
 		"lint": "tslint --project tsconfig.json -t verbose",
 		"postinstall": "node ./node_modules/vscode/bin/install",
 		"test": "node ./node_modules/vscode/bin/test",
-		"antlr4ts": "antlr4ts -visitor ./grammar/mongo.g4 -o ./src"
+		"antlr4ts": "antlr4ts -visitor ./grammar/mongo.g4 -o ./src/mongo/grammar"
```
Comment (PrashanthCorp) [inline package.json]: With a previous refactoring, the generated files now reside in mongo/grammar

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 2]
```diff
@@ -20,13 +20,15 @@ export class MongoCollectionTreeItem implements IAzureParentTreeItem {
...
 	private _cursor: Cursor | undefined;
 	private _hasMoreChildren: boolean = true;
 	private _batchSize: number = DefaultBatchSize;
 
-	constructor(collection: Collection, query?: string) {
+	constructor(collection: Collection, query?: string[]) {
 		this.collection = collection;
-		this._query = query ? JSON.parse(query) : undefined;
+		this._query = query && query.length > 0 ? JSON.parse(query[0]) : undefined;
+		this._projection = query && query.length >= 2 ? JSON.parse(query[1]) : undefined;
```
Comment (PrashanthCorp) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: I wanted to convey a 'length of at least 2' succinctly. Would this be clearer or `query.length > 1` be?

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 3]
```diff
@@ -155,7 +166,17 @@ export class MongoCollectionTreeItem implements IAzureParentTreeItem {
...
+			throw new Error("Too many arguments")
+		}
+		let result;
+		if (args.length === 1) {
+			result = await this.collection.findOne(args[0]);
+		} else if (args.length === 2) {
+			result = await this.collection.findOne(args[0], { fields: args[1] });
+		} else {
+			result = await this.collection.findOne({});
+		}
```
Comment (PrashanthCorp) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: This feels like I'm enumerating function overloading. Any suggestions on a succinct way to do this?

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 4]
```diff
@@ -173,8 +194,21 @@ export class MongoCollectionTreeItem implements IAzureParentTreeItem {
...
+		} else if (args.length === 2) {
+			if (args[1] && args[1].ordered) {
+				collectionWriteOptions["ordered"] = args[1].ordered;
+			}
+			if (args[1] && args[1].writeConcern) {
+				collectionWriteOptions["writeConcern"] = args[1].writeConcern;
+			}
+		}
+
+		return this.collection.insertMany(args[0], collectionWriteOptions)
```
Comment (PrashanthCorp) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: Using `args[0]` feels ugly, but defining a variable for single use seems against practice.

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 5]
```diff
@@ -173,8 +194,21 @@ export class MongoCollectionTreeItem implements IAzureParentTreeItem {
...
+		if (args.length > 2) {
+			throw new Error("Too many arguments. Please see mongo shell documentation. https://docs.mongodb.com/manual/reference/method/db.collection.insertMany/#db.collection.insertMany");
+		} else if (args.length === 2) {
+			if (args[1] && args[1].ordered) {
+				insertManyOptions["ordered"] = args[1].ordered;
+			}
+			if (args[1] && args[1].writeConcern) {
+				insertManyOptions["w"] = args[1].writeConcern;
+			}
+		}
```
Comment (PrashanthCorp) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: The sole-purpose of this is to rename the field _writeConcern_ to _w_

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 6]
```diff
@@ -20,13 +20,15 @@ export class MongoCollectionTreeItem implements IAzureParentTreeItem {
...
+	private readonly _projection: object | undefined;
 	private _cursor: Cursor | undefined;
 	private _hasMoreChildren: boolean = true;
 	private _batchSize: number = DefaultBatchSize;
 
-	constructor(collection: Collection, query?: string) {
+	constructor(collection: Collection, query?: string[]) {
 		this.collection = collection;
-		this._query = query ? JSON.parse(query) : undefined;
+		this._query = query && query.length > 0 ? JSON.parse(query[0]) : undefined;
```
Comment (StephenWeatherford) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: this._query = query && query.length && JSON.parse(query[0]);

same for projection

File: package.json [hunk 7]
```diff
@@ -742,15 +742,15 @@
 		"vsce": "^1.37.5"
 	},
 	"dependencies": {
-		"antlr4ts": "^0.4.0-alpha.4",
+		"antlr4ts": "^0.4.0-burt.2",
```
Comment (StephenWeatherford) [inline package.json]: ?

File: src/mongo/commands.ts [hunk 8]
```diff
@@ -105,7 +105,10 @@ export class MongoScriptDocumentVisitor extends MongoVisitor<MongoCommand[]> {
 		if (argumentsContext) {
 			let functionCallContext = argumentsContext.parent;
 			if (functionCallContext && functionCallContext.parent instanceof mongoParser.CommandContext) {
-				this.commands[this.commands.length - 1].arguments = ctx.text;
+				if (!this.commands[this.commands.length - 1].arguments) {
```
Comment (StephenWeatherford) [inline src/mongo/commands.ts]: extract the expression this.commands[this.commands.length - 1], you're using it three times

File: src/mongo/tree/MongoCollectionTreeItem.ts (hunk 3 above)
Comment (StephenWeatherford) [reply src/mongo/tree/MongoCollectionTreeItem.ts]: This get what you have working and then worry about it.

File: src/mongo/tree/MongoCollectionTreeItem.ts (hunk 5 above)
Comment (StephenWeatherford) [reply src/mongo/tree/MongoCollectionTreeItem.ts]: We'll worry about improvements later

File: package.json (hunk 7 above)
Comment (StephenWeatherford) [reply package.json]: Don't know why it picked that up. Please revert or change to 0.4.1 alpha

image

File: grammar/mongo.g4 [hunk 9]
```diff
@@ -26,7 +35,7 @@ functionCall
 	;
 
 arguments
-	: OPEN_PARENTHESIS = '(' argumentList? CLOSED_PARENTHESIS = ')'
+	: OPEN_PARENTHESIS = '(' (argumentList)* CLOSED_PARENTHESIS = ')'
```
Comment (StephenWeatherford) [inline grammar/mongo.g4]: Where's the comma between arguments?  I can't find it.

File: grammar/mongo.g4 [hunk 10]
```diff
@@ -1,5 +1,14 @@
 grammar mongo;
```
Comment (StephenWeatherford) [inline grammar/mongo.g4]: This gives an error:

db.heroes.insert({  "id": 2,  "name": "Batman",  "saying": "I'm Batman"})

Comment (PrashanthCorp) [reply grammar/mongo.g4]: Thanks! Figured out [the bug](https://github.com/Microsoft/vscode-cosmosdb/pull/460/files/cce9a693f81fcf727762a650d54d0e736cf346ad#diff-3e4e420864699bc495e9daad18f7e036R126).

File: grammar/mongo.g4 [hunk 11]
```diff
@@ -35,7 +35,7 @@ functionCall
 	;
 
 arguments
-	: OPEN_PARENTHESIS = '(' (argumentList)* CLOSED_PARENTHESIS = ')'
+	: OPEN_PARENTHESIS = '(' argumentList (',' argumentList)* CLOSED_PARENTHESIS = ')'
```
Comment (StephenWeatherford) [inline grammar/mongo.g4]: This requires at least one argument, doesn't it?

Comment (StephenWeatherford) [reply grammar/mongo.g4]: Please add a test with empty argument list.
